Add JS replacement for InjectedBundle::FormClient::didAssociateFormControls

achristensen07 opened the pull request:
#### 802028539c2dc706ea8c5b57f0dc98c61a2be57b
<pre>
Add JS replacement for InjectedBundle::FormClient::didAssociateFormControls
<a href="https://bugs.webkit.org/show_bug.cgi?id=297376">https://bugs.webkit.org/show_bug.cgi?id=297376</a>
<a href="https://rdar.apple.com/158244851">rdar://158244851</a>

Reviewed by Ryosuke Niwa.

We introduce a JS event that is dispatched at the same time, but only in content worlds
where _WKContentWorldConfiguration.allowAutofill is YES.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::didAssociateFormControl):
(WebCore::Document::didAssociateFormControlsTimerFired):
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::didAssociateFormControls):
* Source/WebKit/WebProcess/InjectedBundle/API/mac/WKWebProcessPlugInBrowserContextController.mm:
(-[WKWebProcessPlugInBrowserContextController _setFormDelegate:]):
* Source/WebKit/WebProcess/InjectedBundle/APIInjectedBundleFormClient.h:
(API::InjectedBundle::FormClient::didAssociateFormControls):
(API::InjectedBundle::FormClient::shouldNotifyOnFormChanges):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFormClient.cpp:
(WebKit::InjectedBundlePageFormClient::didAssociateFormControls):
(WebKit::InjectedBundlePageFormClient::shouldNotifyOnFormChanges):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFormClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::didAssociateFormControls):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Tools/TestWebKitAPI/Tests/WebKit/DidAssociateFormControls.cpp:
(TestWebKitAPI::TEST(WebKit, DidAssociateFormControlsWithInjectedBundle)):
(TestWebKitAPI::TEST(WebKit, DidAssociateFormControls)): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FormValidation.mm:
(TestWebKitAPI::TEST(FormValidation, FormValidationOnUnparentedWindowDoesNotCrash)):
(TestWebKitAPI::TEST(WebKit, DidAssociateFormControls)):
(TEST(FormValidation, PresentingFormValidationUIWithoutViewControllerDoesNotCrash)): Deleted.
(TEST(FormValidation, FormValidationOnUnparentedWindowDoesNotCrash)): Deleted.

Canonical link: <a href="https://commits.webkit.org/298736@main">https://commits.webkit.org/298736@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b3797d4fc1618d27eefca086d4b200f69309ec9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116500 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36161 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26713 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122557 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/67063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118389 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36855 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44749 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/88469 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/67063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119449 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29388 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104507 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/68913 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28456 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22616 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66238 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98763 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22774 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125706 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43394 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32584 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/97174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43759 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100716 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/96969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42265 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20175 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/39348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18601 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43282 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48873 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42748 "Built successfully") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/46088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44453 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

arenevier commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9820,8 +9820,14 @@ CompositeOperator Document::compositeOperatorForBackgroundColor(const Color& col
+    if (!page)
         return;
+    if (!page->chrome().client().shouldNotifyOnFormChanges()) {
+        // Is it sufficient to only support registering event listeners on the window object, or do we need to support other event targets?
```
For autofill, and more generally for user scripts in safari, we add event listener on the window. So it's fine.

rniwa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9820,8 +9820,14 @@ CompositeOperator Document::compositeOperatorForBackgroundColor(const Color& col
+    if (!page->chrome().client().shouldNotifyOnFormChanges()) {
+        // Is it sufficient to only support registering event listeners on the window object, or do we need to support other event targets?
+        // Is this check fast enough to prevent perf slowdown?
+        if (m_domWindow && !m_domWindow->hasEventListeners(eventNames().webkitassociateformcontrolsEvent))
```
You wanna check `document.hasEventListenersOfType` instead. That map will contain the event listeners throughout the document, not just the ones registered at m_domWindow.

rniwa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/FormValidation.mm:
```diff
@@ -54,9 +60,30 @@
+    RetainPtr configuration = adoptNS([_WKContentWorldConfiguration new]);
+    configuration.get().allowAutofill = YES;
+    RetainPtr autofillWorld = [WKContentWorld _worldWithConfiguration:configuration.get()];
+    NSString *pageWorldJS = @"window.addEventListener('webkitassociateformcontrols', ()=>{ alert('fail') })";
```
You don't need { } around alert. You can just say `() => alert('fail')`.

rniwa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/FormValidation.mm:
```diff
@@ -54,9 +60,30 @@
+    configuration.get().allowAutofill = YES;
+    RetainPtr autofillWorld = [WKContentWorld _worldWithConfiguration:configuration.get()];
+    NSString *pageWorldJS = @"window.addEventListener('webkitassociateformcontrols', ()=>{ alert('fail') })";
+    NSString *autofillWorldJS = @"window.addEventListener('webkitassociateformcontrols', ()=>{ setTimeout(()=>{ alert('pass') }, 100)})";
```
Ditto for the inner lambda.

rniwa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9839,6 +9845,13 @@ void Document::didAssociateFormControlsTimerFired()
+        Ref event = Event::create(eventNames().webkitassociateformcontrolsEvent, Event::CanBubble::No, Event::IsCancelable::No);
+        event->setIsAutofillEvent();
+        // Do we need to put all the controls into the event somehow?
+        // Should we make a new Event idl and class for this list of elements?
```
We most definitely don't want to dispatch an event with a list of elements.
That's just generally not how an event works.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/590d1b307aeb13dfe9d229777136a664943ea570. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49374

arenevier commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9822,6 +9822,8 @@ CompositeOperator Document::compositeOperatorForBackgroundColor(const Color& col
 void Document::didAssociateFormControl(Element& element)
 {
     RefPtr page = this->page();
+    if (!page)
```
This check is redundant with the check on next line.

rniwa commented on Source/WebKit/WebProcess/InjectedBundle/API/mac/WKWebProcessPlugInBrowserContextController.mm:
```diff
@@ -546,15 +546,15 @@ bool shouldNotifyOnFormChanges(WebKit::WebPage*) final
             auto controller = m_controller.get();
             if (!controller)
-                return false;
+                return true;
```
Why are we changing these return values?

rniwa replied:
Oh I guess you want to make Document::didAssociateFormControl schedule a timer.
I think we should add a new check for document.hasEventListenersOfType instead.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fbfd1880aba4afa9a8108f72ca88d39fb7eebcc2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49374

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6b3797d4fc1618d27eefca086d4b200f69309ec9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49374

achristensen07 committed:
Add JS replacement for InjectedBundle::FormClient::didAssociateFormControls
https://bugs.webkit.org/show_bug.cgi?id=297376
rdar://158244851

Reviewed by Ryosuke Niwa.

We introduce a JS event that is dispatched at the same time, but only in content worlds
where _WKContentWorldConfiguration.allowAutofill is YES.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::didAssociateFormControl):
(WebCore::Document::didAssociateFormControlsTimerFired):
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::didAssociateFormControls):
* Source/WebKit/WebProcess/InjectedBundle/API/mac/WKWebProcessPlugInBrowserContextController.mm:
(-[WKWebProcessPlugInBrowserContextController _setFormDelegate:]):
* Source/WebKit/WebProcess/InjectedBundle/APIInjectedBundleFormClient.h:
(API::InjectedBundle::FormClient::didAssociateFormControls):
(API::InjectedBundle::FormClient::shouldNotifyOnFormChanges):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFormClient.cpp:
(WebKit::InjectedBundlePageFormClient::didAssociateFormControls):
(WebKit::InjectedBundlePageFormClient::shouldNotifyOnFormChanges):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFormClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::didAssociateFormControls):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Tools/TestWebKitAPI/Tests/WebKit/DidAssociateFormControls.cpp:
(TestWebKitAPI::TEST(WebKit, DidAssociateFormControlsWithInjectedBundle)):
(TestWebKitAPI::TEST(WebKit, DidAssociateFormControls)): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FormValidation.mm:
(TestWebKitAPI::TEST(FormValidation, FormValidationOnUnparentedWindowDoesNotCrash)):
(TestWebKitAPI::TEST(WebKit, DidAssociateFormControls)):
(TEST(FormValidation, PresentingFormValidationUIWithoutViewControllerDoesNotCrash)): Deleted.
(TEST(FormValidation, FormValidationOnUnparentedWindowDoesNotCrash)): Deleted.

Canonical link: https://commits.webkit.org/298736@main
